workers/unity/Packages/com.improbable.gdk.mobile/Editor/LaunchMenu.cs

```diff
@@ -70,25 +75,90 @@ private static void LaunchMobileClient()
                     .WithArgs("shell", "am", "start", "-S", "-n", $"{bundleId}/com.unity3d.player.UnityPlayerActivity",
                         "-e", "\"arguments\"", $"\\\"{arguments.ToString()}\\\"")
                     .Run();
-
-                EditorUtility.DisplayProgressBar("Launching Mobile Client", "Done", 1.0f);
             }
             finally
             {
                 EditorUtility.ClearProgressBar();
             }
         }
 
-        private static bool TryGetApkPath(string rootPath, out string apkPath)
+        [MenuItem(MenuLaunchiOSDevice, false, 74)]
+        private static void LaunchiOSDeviceClient()
         {
-            foreach (var file in Directory.GetFiles(rootPath, "*.apk", SearchOption.AllDirectories))
+            try
             {
-                apkPath = file;
-                return true;
-            }
+                // Ensure needed tools are installed
+                if (string.IsNullOrEmpty(LibIDeviceInstallerBinary))
+                {
+                    Debug.LogError("Could not find ideviceinstaller tool. Please ensure it is installed. " +
+                        "See https://github.com/libimobiledevice/ideviceinstaller fore more details.");
+                    return;
+                }
+
+                if (string.IsNullOrEmpty(LibIDeviceDebugBinary))
+                {
+                    Debug.LogError("Could not find idevicedebug tool. Please ensure libimobiledevice is installed. " +
+                        "See https://helpmanual.io/help/idevicedebug/ for more details.");
+                    return;
+                }
+
+                EditorUtility.DisplayProgressBar("Launching iOS Device Client", "Installing archive", 0.3f);
+
+                // Get chosen ios package id
+                var bundleId = PlayerSettings.GetApplicationIdentifier(BuildTargetGroup.iOS);
+                string appList = null;
+
+                // Find archive to install
+                var ipaPath = Directory.GetFiles(AbsoluteAppBuildPath, "*.ipa", SearchOption.AllDirectories).FirstOrDefault();
+                RedirectedProcess.Command(LibIDeviceInstallerBinary).WithArgs("-l")
+                    .AddAccumulatedOutputProcessing(output => { appList = output; })
+                    .RedirectOutputOptions(OutputRedirectBehaviour.None).Run();
+                var existsOnDevice = appList.Contains(bundleId);
+                var existsLocally = !string.IsNullOrEmpty(ipaPath);
+                if (!existsLocally && !existsOnDevice)
```

Comment:
would separate that out into two check:
first: check if ipaPath exists: if not, print an error directly related to that and mention that you first need to build it
second: check if the app exists on the device: if not, print an errror mentioning the installation has failed. 

Is it possible to retrieve the reason of why it could have failed?

Reply:
The logic here is following:
If .ipa exists - we install/upgrade device app and launch it
If .ipa doesn't exist, but the bunlde is installed on device - we launch existing app with parameters
If .ipa doesn't exist and bundle isn't found on device - we fail with an error.

Let me add a comment in code explaining it.

All important output is redirected to the console, so that it is visible to developer.